io/io/src/TStreamerInfoActions.cxx

```diff
@@ -1609,405 +1620,447 @@ namespace TStreamerInfoActions
+               Looper::template LoopOverCollection< ReadStreamerLoopStatic > (buf, start, args..., config);
+            } // fileVersion
+            // Backpatch the byte count into the buffer.
+            buf.SetByteCount(pos, kTRUE);
 
-            // } // for k
-            buf.CheckByteCount(start, count, config->fCompInfo->fElem->GetFullName());
-            // We are done, next streamer element.
             return 0;
          }
+      };
 
+      template<bool kIsTextT>
+      static Int_t ReadStreamerLoopPoly(TBuffer &buf, void *addr, const TConfiguration *config) {
+         // Get the class of the data member.
+         TClass* cl = config->fCompInfo->fClass;
+         UInt_t ioffset = config->fOffset;
          // Which are we, an array of objects or an array of pointers to objects?
-         Bool_t isPtrPtr = (strstr(config->fCompInfo->fElem->GetTypeName(), "**") != 0);
-
-         // By default assume the file version is the newest.
-         Int_t fileVersion = kMaxInt;
-         if (!kIsTextT) {
-            // At this point we do *not* have a private streamer.
-            // Get the version of the file we are reading from.
-            TFile* file = (TFile*) buf.GetParent();
-            if (file) {
-               fileVersion = file->GetVersion();
-            }
-         }
-         // Read the class version and byte count from the buffer.
-         UInt_t start = 0;
-         UInt_t count = 0;
-         buf.ReadVersion(&start, &count, config->fInfo->IsA());
-         if (fileVersion > 51508) {
-            // -- Newer versions allow polymorphic pointers.
-            // Loop over the entries in the clones array or the STL container.
-            // for (Int_t k = 0; k < narr; ++k) {
-            // Get the counter for the varying length array.
-            Int_t vlen = *((Int_t *)((char *)addr /*entry pointer*/ + eoffset /*entry offset*/ +
-                                    config->fCompInfo->fMethod /*counter offset*/));
-            // Int_t realLen;
-            // b >> realLen;
-            // if (realLen != vlen) {
-            //   fprintf(stderr, "read vlen: %d  realLen: %s\n", vlen, realLen);
-            //}
-            // Get a pointer to the array of pointers.
-            char **pp = (char **)((char *)addr /*entry pointer*/ + ioffset /*object offset*/);
-            // Loop over each element of the array of pointers to varying-length arrays.
-            // if (!pp) {
-            //   continue;
-            // }
+         bool isPtrPtr = ((TConfStreamerLoop*)config)->fIsPtrPtr;
+
+         // Get the counter for the varying length array.
+         Int_t vlen = *((Int_t *)((char *)addr /*entry pointer*/ +
+                                 config->fCompInfo->fMethod /*counter offset*/));
+         // Int_t realLen;
+         // b >> realLen;
+         // if (realLen != vlen) {
+         //   fprintf(stderr, "read vlen: %d  realLen: %s\n", vlen, realLen);
+         //}
```

Comment:
Here and further down: remove code that's commented out or add comment why it should stay?